### ISSUE TITLE
Partial application for built-in functions.

### ISSUE THREAD

**Issue by kwxm** [input-output-hk/plutus#1144]
Some time at the end of 2018 we decided that built-in functions should always be fully applied, for reasons that I don't recollect.

I think @michaelpj and @effectfully are keen to drop this and go back to partial application. 

I quite like having builtins be fully applied, for the following reason.  In semantics and type theory we usually pretend that all functions have only one argument because it simplifies presentation and reasoning, but it's a really bad idea to do this in an implementation because then you have to create intermediate closures that are immediately eliminated: you write `f 1 2 3` but it's converted to `((f 1) 2) 3` and you have to create a closure for `f 1` and then another one for `(f 1) 2`, and eliminate them both when you get the 3 in order to calculate the final result, wasting time and space (in the CK/CEK machines, you're not actually creating closures, but saving frames on the stack of execution contexts, but it's essentially the same thing).  There are two ways to get round this: either do some complicated optimisation during compilation/execution (which is what most functional languages do), or demand that everything's fully applied in the first place.  Another possibility is just to put up with the inefficiency, and maybe that is what we should do here because it keeps things simpler.

Sticking to full application for builtins would certainly make evaluation of_builtins_ more efficient, but we're definitely stuck with single application for non-builtin functions, and maybe that will drown out any gains to be made from efficient builtin application.  Ultimately I'm on the fence about this.

I think we should

* Try to remember why we switched to full application for builtins in the first place.
* Hear the arguments for reverting to partial application.
* Have a discussion and make a decision.

**Comment by effectfully** [comment]
I remember Philip saying that the main thing about saturated builtins is not that they are faster, but that they are more correct (or something like that). I don't understand at the moment what is more correct about saturated builtins.

> Sticking to full application for builtins would certainly make evaluation of_builtins_ more efficient, but we're definitely stuck with single application for non-builtin functions, and maybe that will drown out any gains to be made from efficient builtin application.

This is a very good point.

**Comment by michaelpj** [comment]
In particular, if we have saturated builtins then the Plutus Tx compiler will definitely need to wrap all the builtins in other functions (i.e. `let addInteger = \x y -> addInteger x y in ...`) because the Haskell-level functions that we convert to builtins need not have fully saturated applications.

So we may not gain that much efficiency in practice. It would probably make the evaluator simpler, though.

I agree with @effectfully that there was *something* about it making the theory simpler, but I'm not sure what.

Here's one thing that's awkward: you can't tell whether a builtin application will reduce or not without looking at the whole spine of arguments. e.g. if we look at `(((\x y . 1) t) u)` we can tell when we get to the inner application that it will reduce, but we couldn't do that if it was a builtin until we've seen all the arguments.

Overall I have moved to being ambivalent about which way we should go.

**Comment by kwxm** [comment]
I had a question about type quantification for builtins, but now I'm confused.

Currently the builtin functions all have signatures of the form
```
  forall a_1::K_1 ... forall a_m::K_m . t_1 -> t_2 -> .... -> t_n -> u
```
What I was worried about is that when everything's fully appplied we instantiate all of the type variables at the start, before we start applying the function to term arguments.  If we allow partial application of builtins then there's a possiblity that we might want to allow partial type instantiation too, so that we'd have quantifiers appearing in the middle of the signature as well.  I'm wondering if we'd ever want to do that in practice, since it'd involve changing the syntax of builtin application to allow type instantiation part-way through.  We're already debating the concrete syntax of builtin application in #1143, so this would affect that as well.

However, I now notice that in the signature above there's nothing to stop some `t_i` from being a universally quantified type anyway, and we don't have any way of instantiating the type.  Does that matter?  As I say, I've just managed to confuse myself.

**Comment by michaelpj** [comment]
Why is partial instantiation any more problematic than partial application? And given the signatures of builtins all the instantiations would have to come first.

The "partial application is okay" approach is just that a builtin *is* a function of signature `forall a1...am . t1 -> ... -> tn -> u`, and you can do with that exactly what you could with any other function.

I don't think we're going to want higher-rank builtins (he says optimistically), so we shouldn't need quantifiers in the middle. So I agree that the `t` types should be quantifier-free.

**Comment by kwxm** [comment]
> I don't think we're going to want higher-rank builtins (he says optimistically), so we shouldn't need quantifiers in the middle. So I agree that the `t` types should be quantifier-free.

Good!  I was hoping that was the position, but you never know what people might want to do...

**Comment by michaelpj** [comment]
I mean, maybe we'll want to offer a builtin implementation of `ST` that's guaranteed to be fast ;)

**Comment by kwxm** [comment]
We should decide what to do here.  Maybe now that we have more experience it'll be clearer whether having unsaturated builtins would make things any better.

**Comment by michaelpj** [comment]
@james can you just confirm what the metatheory does and whether there are any metatheoretic considerations?

**Comment by effectfully** [comment]
I'm against it. In Agda all your types are rich and any Agda programmer is used to it. But making a core AST richly typed in Haskell is not as nice. It also doesn't play well with `deriving`. And keeping core data types simple is generally a good idea.

The formalization also hardcodes a bunch of builtins (as well as the implementation), but that is not what we want to do with the implementation. We want builtins to be extensible. Which seems to entail that the AST has to be richly typed if we are to enforce saturatedness.

I haven't heard any appealing arguments in favor of saturated application for builtins: the efficiency argument doesn't sound too convincing, because regular function application stays unoptimized anyway and because we didn't really optimize anything, so why optimize only builtins. We could get away without reflecting the whole type signature of a builtin, but at least we'd have to reflect the number of arguments that it receives, I think. But given that we reflect the type signatures for evaluation purposes, we'd have to ensure that a reflected type signature agrees with the reflected number of arguments and while I think it's manageable, it certainly adds complexity to the codebase. 

But if someone wants to try out saturated builtins and see whether there is any fine way to get them, go ahead. I just wanted to point out that if they're implemented in the metatheory nicely, that doesn't really mean anything in the Haskell setting. But metatheoretic considerations are important of course, so if there are any, then I'd also like to hear them.

**Comment by kwxm** [comment]
> @james can you just confirm what the metatheory does and whether there are any metatheoretic considerations?

I think you want @jmchapman on GitHub.  There is a 'james', but he's someone completely different.  Life was so much simpler when people only had one identity.

**Comment by kwxm** [comment]
> the efficiency argument doesn't sound too convincing, because regular function application stays unoptimized anyway and because we didn't really optimize anything, so why optimize only builtins. 

That's my suspicion as well, but it's something we could actually measure if we revert to unsaturated builtins.

**Comment by michaelpj** [comment]
Yikes, apologies to other James :grimacing:

**Comment by jmchapman** [comment]
The intrinsically typed formalisation sticks very rigidly to the saturated approach of the spec - unsatuated builtins are unrepresentable. The extrinsically typed formalisation that I use for testing evaluation allows unsaturated builtins in a rather adhoc manner.

**Comment by kwxm** [comment]
> ... it's something we could actually measure if we revert to unsaturated builtins. 

Having thought a bit more, that's a pointless remark.  The current CEK implementation doesn't deal efficiently with saturated builtin application.  If I recall corectly, it just consumes more and more arguments until application is successful: if you want to evaluate `b x y z` it first tries `b x` and gets a failure, then tries `b x y` and gets another one, then finally tries `b x y z` and succeeds.  The machine doesn't have any information about arities of built-in functions, so it doesn't know in advance how many arguments it needs to supply.

**Comment by michaelpj** [comment]
> Having thought a bit more, that's a pointless remark. The current CEK implementation doesn't deal efficiently with saturated builtin application.

> ... but modifying that strategy to handle unsaturated builtins might not be too hard.

The current Haskell implementation has unsaturated builtins and I think they work just fine! As you say, the evaluator just consumes arguments until the application is saturated, otherwise it's stuck.

To recapitulate:
- The Haskell implementation has unsaturated builtins.
- The spec and Agda implementation have saturated builtins.

One of these parties needs to change, it's just a question of which.

**Comment by kwxm** [comment]
@effectfully says he'll be able to put some time into working on extensible builtins (#581) soon, so that may help to settle this.

**Comment by effectfully** [comment]
> The formalization also hardcodes a bunch of builtins (as well as the implementation), but that is not what we want to do with the implementation. We want builtins to be extensible. Which seems to entail that the AST has to be richly typed if we are to enforce saturatedness.

I don't think we should enforce saturatedess (which is what the spec and the formalization do IIRC). We can just change the `Builtin` `Term` constructor to store a `Builtin`, a list of `Type` arguments and a (non-empty?) list of `Term` arguments. I think we should try doing that.

**Comment by michaelpj** [comment]
Yeah, I've come around to saturated builtins, I think it's not so bad.

**Comment by michaelpj** [comment]
Roman, when you say we shouldn't enforce saturatedness, do you mean that the typechecker shouldn't check these "applied builtin" nodes for full saturation? I'm not sure what the point of having them is if we're not requiring them to be saturated.

**Comment by effectfully** [comment]
> Roman, when you say we shouldn't enforce saturatedness, do you mean that the typechecker shouldn't check these "applied builtin" nodes for full saturation? I'm not sure what the point of having them is if we're not requiring them to be saturated.

No, the type checker definitely should check that. What I don't want is enforcing saturatedness at the type level, which is what the spec and the formalization do (IIRC). This was the main reason why I was opposed to implementing what's in the spec.

**Comment by michaelpj** [comment]
Yes, I agree, our AST type shouldn't enforce well-formedness of builtins.

I think what we're suggesting is:
- An "applied builtin" AST node, replacing the current one.
    - These "should be" saturated
- The typechecker checks that applied builtins are saturated
- The evaluator assumes they're saturated and throws an error if they're not
